test/e2e/functional/tests/e2e.go

```diff
@@ -73,148 +76,241 @@ var _ = Describe("metallb", func() {
 					}
 
-					for _, pod := range pods.Items {
-						if pod.Status.Phase != corev1.PodRunning {
-							return fmt.Errorf("deployment %s pod %s is not running, expected status %s got %s", consts.MetalLBOperatorDeploymentName, pod.Name, corev1.PodRunning, pod.Status.Phase)
+					EventuallyWithOffset(1, func() []corev1.Container {
+						speakerDaemonSet := &appsv1.DaemonSet{}
+						err := testclient.Client.Get(
+							context.Background(),
+							types.NamespacedName{Name: consts.MetalLBDaemonsetName, Namespace: metallb.Namespace},
+							speakerDaemonSet)
+						if err != nil {
+							return nil
 						}
-					}
 
-					return nil
-				}, metallbutils.DeployTimeout, metallbutils.Interval).ShouldNot(HaveOccurred())
-			})
+						return speakerDaemonSet.Spec.Template.Spec.Containers
+					}, 2*time.Second, 200*time.Millisecond).Should(
+						ContainElement(
+							And(
+								WithTransform(nameGetter, Equal("speaker")),
+								WithTransform(envGetter, bgpTypeMatcher),
+							)))
+				}
 
-			By("checking MetalLB daemonset is in running state", func() {
-				Eventually(func() error {
-					daemonset, err := testclient.Client.DaemonSets(metallb.Namespace).Get(context.Background(), consts.MetalLBDaemonsetName, metav1.GetOptions{})
-					if err != nil {
-						return err
-					}
+				By("setting the bgpType")
 
-					pods, err := testclient.Client.Pods(OperatorNameSpace).List(context.Background(), metav1.ListOptions{
-						LabelSelector: "component=speaker"})
-					if err != nil {
-						return err
-					}
+				err := testclient.Client.Get(context.Background(), goclient.ObjectKey{Namespace: metallb.Namespace, Name: metallb.Name}, metallb)
+				if !errors.IsNotFound(err) {
+					Expect(err).ToNot(HaveOccurred())
+				}
 
-					if len(pods.Items) != int(daemonset.Status.DesiredNumberScheduled) {
-						return fmt.Errorf("daemonset %s pods are not ready, expected %d generations got %d pods", consts.MetalLBDaemonsetName, int(daemonset.Status.DesiredNumberScheduled), len(pods.Items))
-					}
+				if errors.IsNotFound(err) {
+					metallb := &metallbv1beta1.MetalLB{
+						ObjectMeta: metav1.ObjectMeta{
+							Name:      metallb.Name,
+							Namespace: metallb.Namespace,
+						},
+						Spec: metallbv1beta1.MetalLBSpec{
+							LogLevel:   metallbv1beta1.LogLevelWarn,
+							BGPBackend: bgpType,
+						},
+					}
+					err := testclient.Client.Create(context.Background(), metallb)
+					Expect(err).ToNot(HaveOccurred())
+				} else {
+					metallb.Spec.BGPBackend = bgpType
+					err = testclient.Client.Update(context.Background(), metallb)
+					Expect(err).ToNot(HaveOccurred())
+				}
 
-					for _, pod := range pods.Items {
-						if pod.Status.Phase != corev1.PodRunning {
-							return fmt.Errorf("daemonset %s pod %s is not running, expected status %s got %s", consts.MetalLBDaemonsetName, pod.Name, corev1.PodRunning, pod.Status.Phase)
+				By("checking the controller is running in the right bgp mode", func() {
```

Comment:
mind using By(desc) instead By(desc,func())? the extra indents don't help imo 😅

Reply:
done